### PR TITLE
[SPARK-2475][Streaming] Check whether has enough resources to process the received data in local mode

### PR DESCRIPTION
**Pull request by ArcherShao**

### CODE REVIEW

**Comment by ArcherShao** [comment]
@tdas can you take a review on this? thanks.

File: streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala [hunk 1]
```diff
@@ -488,6 +488,8 @@ class StreamingContext private[streaming] (
     assert(graph != null, "Graph is null")
     graph.validate()
 
+    assert(isCoresEnoughInLocalMode(), "No enough cores, Spark jobs will not get resources to " +
```
Comment (huitseeker) [inline streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala]: You mean 'Not' enough cores.

File: streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala [hunk 2]
```diff
@@ -495,6 +497,45 @@ class StreamingContext private[streaming] (
...
+    val LOCAL_N_FAILURES_REGEX = """local\[([0-9]+|\*)\s*,\s*([0-9]+)\]""".r
+    // Regular expression for simulating a Spark cluster of [N, cores, memory] locally
+    val LOCAL_CLUSTER_REGEX = """local-cluster\[\s*([0-9]+)\s*,\s*([0-9]+)\s*,\s*([0-9]+)\s*]""".r
+
+    val master = sc.conf.get("spark.master")
+    val coresPerTask = sc.conf.getInt("spark.task.cpus", 1)
+    val maxTaskNum = master match {
+      case "local" => 1 / coresPerTask
+
+      case LOCAL_N_REGEX(threads) =>
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala]: This effectively replicates the logic in SparkContext, DAGScheduler, etc. This is not good. Things can silently break if any of the internal logic of Spark change. A better solution might be to hook into the Spark's internal DAGScheduler, etc to see the number of cores it chooses.

**Comment by tdas** [comment]
Do you have more time to work on this? If not, mind closing this PR?

**Pull request closed without merging**